lib/optimizely/odp/lru_cache.rb

```diff
@@ -0,0 +1,116 @@
+# frozen_string_literal: true
+
+#
+#    Copyright 2022, Optimizely and contributors
+#
+#    Licensed under the Apache License, Version 2.0 (the "License");
+#    you may not use this file except in compliance with the License.
+#    You may obtain a copy of the License at
+#
+#        http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+#
+
+module Optimizely
+  class LRUCache
+    # Least Recently Used cache that invalidates entries older than the timeout.
+
+    attr_reader :capacity, :timeout
+
+    def initialize(capacity, timeout_in_secs)
+      # @param capacity - The max size of the cache. If set <= 0, caching is disabled.
+      # @param timeout_in_secs - Seconds until a cache item is considered stale.
+      #                          If set <= 0, items never expire.
+      @cache_mutex = Mutex.new
+      @map = {}
+      @capacity = capacity
+      @timeout = timeout_in_secs
+    end
+
+    # Retrieve the non stale value from the cache corresponding to the provided key
+    # or nil if key is not found
+    # Moves the key/value pair to the end of the cache
+    #
+    # @param key - The key to retrieve
+
+    def lookup(key)
+      return nil if @capacity <= 0
+
+      @cache_mutex.synchronize do
+        return nil unless @map.include?(key)
+
+        old_element = @map[key]
+        @map.delete(key)
+        return nil if old_element.stale?(@timeout)
+
+        element = CacheElement.new(old_element.value)
+        @map[key] = element
```

Comment:
It looks like ruby hash keeps insertion order for 1.9+. Min ruby version we support?

Reply:
That's one of the main reasons I pushed for updating the minimum supported ruby version. As of 4.0.0, we now only support >=2.7

Reply:
@andrewleap-optimizely That min version info should be in README.md file and our docs. Updated?

Reply:
The minimum version is listed on the rubygems page but yea we should add it to the readme as well. Looks like a lot of the sdks are missing that information on their readme/docs, may be worth discussing.